tests/contrib/urllib3/test_urllib3.py

```diff
@@ -0,0 +1,392 @@
+
+
+class BaseUrllib3TestCase(object):
+    """Provides the setup and teardown for patching/unpatching the urllib3 integration"""
+
+    def setUp(self):
+        super(BaseUrllib3TestCase, self).setUp()
+
+        patch()
+        self.http = urllib3.PoolManager()
+        setattr(urllib3.connectionpool.HTTPConnectionPool, "datadog_tracer", self.tracer)
+
+    def tearDown(self):
+        unpatch()
+
+        super(BaseUrllib3TestCase, self).tearDown()
+
+
+class TestUrllib3(BaseUrllib3TestCase, BaseTracerTestCase):
+    def test_HTTPConnectionPool_traced(self):
+        """Tests that requests made from the HTTPConnectionPool are traced"""
+        pool = urllib3.connectionpool.HTTPConnectionPool(SOCKET)
+        # Test a relative URL
+        r = pool.request("GET", "/status/200")
+        assert r.status == 200
+        assert len(self.tracer.writer.pop()) == 1
+
+        # Test an absolute URL
+        r = pool.request("GET", URL_200)
+        assert r.status == 200
+        assert len(self.tracer.writer.pop()) == 1
+
+    def test_traced_connection_from_url(self):
+        """Tests tracing from ``connection_from_url`` is set up"""
+        conn = urllib3.connectionpool.connection_from_url(URL_200)
+        resp = conn.request("GET", "/")
+        assert resp.status == 200
+        assert len(self.tracer.writer.pop()) == 1
+
+    def test_resource_path(self):
+        """Tests that a successful request tags a single span with the URL"""
+        resp = self.http.request("GET", URL_200)
+        assert resp.status == 200
+        spans = self.tracer.writer.pop()
+        assert len(spans) == 1
+        s = spans[0]
+        assert s.get_tag("http.url") == URL_200
+
+    def test_tracer_disabled(self):
+        """Tests a disabled tracer produces no spans on request"""
+        self.tracer.enabled = False
+        out = self.http.request("GET", URL_200)
+        assert out.status == 200
+        spans = self.tracer.writer.pop()
+        assert len(spans) == 0
+
+    def test_args_kwargs(self):
+        """
+        Test that args are kwargs are correctly inferred from the target function's
+        signature.
+
+        The args/kwargs used in the integration are:
+            - method (idx 0)
+            - url (idx 1)
+            - headers (idx 3)
+        """
+        inputs = [
+            (("POST", URL_200, b"payload", {"accept": "*"}), {}),
+            (("POST", URL_200, b"payload"), {"headers": {"accept": "*"}}),
+            (("POST", URL_200), {"headers": {"accept": "*"}}),
+            (("POST",), {"url": URL_200, "headers": {"accept": "*"}}),
+            ((), {"method": "POST", "url": URL_200, "headers": {"accept": "*"}}),
+        ]
+
+        for args, kwargs in inputs:
+
+            with self.override_config("urllib3", {}):
+                config.urllib3.http.trace_headers(["accept"])
+                pool = urllib3.connectionpool.HTTPConnectionPool(SOCKET)
+                out = pool.urlopen(*args, **kwargs)
```

Comment:
In the above tests you are calling `pool.request` but here `urlopen`.

```suggestion
                out = pool.request(*args, **kwargs)
```

Reply:
Specifically this test is checking the signature is correctly interpreted by the patch function `urlopen` is correct sing that is the function being patched here. Tests should be sure to cover all ways in which the user will use the API, which is why some of the tests at the top are using different classes and functions.